demo_nodes_cpp/src/parameters/parameter_events_async.cpp

```diff
@@ -15,71 +15,123 @@
 #include <chrono>
 #include <iostream>
 #include <memory>
+#include <vector>
 
 #include "rclcpp/rclcpp.hpp"
 
 using namespace std::chrono_literals;
+using SetParametersResult =
+    std::shared_future<std::vector<rcl_interfaces::msg::SetParametersResult>>;
 
-void on_parameter_event(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+class ParameterEventsAsyncNode : public rclcpp::Node
 {
-  // TODO(wjwwood): The message should have an operator<<, which would replace all of this.
-  std::cout << "Parameter event:" << std::endl << " new parameters:" << std::endl;
-  for (auto & new_parameter : event->new_parameters) {
-    std::cout << "  " << new_parameter.name << std::endl;
+public:
+  ParameterEventsAsyncNode()
+  : Node("parameter_events")
+  {
+    // Typically a parameter client is created for a remote node by passing the name of the remote
+    // node in the constructor; in this example we create a parameter client for this node itself.
+    parameters_client_ = std::make_shared<rclcpp::AsyncParametersClient>(this);
+
+    auto on_parameter_event_callback =
+      [](const rcl_interfaces::msg::ParameterEvent::SharedPtr event) -> void
+      {
+        // TODO(wjwwood): The message should have an operator<<, which would replace all of this.
+        std::cout << "Parameter event:" << std::endl << " new parameters:" << std::endl;
+        for (auto & new_parameter : event->new_parameters) {
+          std::cout << "  " << new_parameter.name << std::endl;
+        }
+        std::cout << " changed parameters:" << std::endl;
+        for (auto & changed_parameter : event->changed_parameters) {
+          std::cout << "  " << changed_parameter.name << std::endl;
+        }
+        std::cout << " deleted parameters:" << std::endl;
+        for (auto & deleted_parameter : event->deleted_parameters) {
+          std::cout << "  " << deleted_parameter.name << std::endl;
+        }
+      };
+
+    // Setup callback for changes to parameters.
+    parameter_event_sub_ = parameters_client_->on_parameter_event(on_parameter_event_callback);
+
+    // Queue a `set_parameters` request as soon as `spin` is called on this node.
+    timer_ = create_wall_timer(0s,
+        [this]() {
+          this->queue_first_set_parameter_request();
+        });
   }
-  std::cout << " changed parameters:" << std::endl;
-  for (auto & changed_parameter : event->changed_parameters) {
-    std::cout << "  " << changed_parameter.name << std::endl;
+
+  // Set several different types of parameters.
+  void queue_first_set_parameter_request()
+  {
+    timer_->cancel();
```

Comment:
I think it would be nice to have a comment about why this cancel is here, i.e. "cancel the timer to prevent it from running more than once".

Also, it might be worth adding a TODO to make a way to do so called "one shot timers" in the rclcpp api, so this wouldn't be necessary. We could replace the `create_wall_timer` thing above with a `call_soon(<function>)` or if you really wanted a timer, have a parameter to `create_wall_timer` like `bool one_shot=false`.